File: ts-multi-region-s3-crr-kms-cmk/multi-region-s3-crr-kms-cmk/lib/multi-region-s3-crr-kms-cmk-stage.ts
import { Construct } from 'constructs';
import {Stage, StageProps} from 'aws-cdk-lib';
import { MultiRegionS3CrrKmsCmkTargetStack } from './multi-region-s3-crr-kms-cmk-target-stack';
import { MultiRegionS3CrrKmsCmkSourceStack } from './multi-region-s3-crr-kms-cmk-source-stack';

export class MultiRegionS3CrrKmsCmkStage extends Stage {
    constructor(scope: Construct, id: string, props?: StageProps) {
        super(scope, id, props);

        const targetStack = new MultiRegionS3CrrKmsCmkTargetStack(this, 'MultiRegionS3CrrKmsCmkTarget', {
            env: { account: props?.env?.account, region: "us-west-2"}
        });

        const sourceStack = new MultiRegionS3CrrKmsCmkSourceStack(this, 'MultiRegionS3CrrKmsCmkSource', {
            env: { account: props?.env?.account, region: 'us-west-1' },
            targetBucket: targetStack.targetBucket,
            targetKeyIdSsmParameterName: targetStack.targetKeyIdSsmParameterName,
            targetRegion: targetStack.region
        });

        sourceStack.addDependency(targetStack);
    }
}

File: go-multi-region-s3-crr-kms-cmk/multi-region-s3-crr-kms-cmk/internal/pipeline/multi-region-s3-crr-kms-cmk-stage.go
package pipeline

import (
	"multi-region-s3-crr-kms-cmk/internal/stacks"

	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

func NewMultiRegionS3CrrKmsCmkStage(scope constructs.Construct, id *string, props awscdk.StageProps) awscdk.Stage {
	stage := awscdk.NewStage(scope, id, &props)

	targetStack := stacks.NewMultiRegionS3CrrKmsCmkTargetStack(stage, jsii.String("MultiRegionS3CrrKmsCmkTarget"), &awscdk.StackProps{
		Env: &awscdk.Environment{
			Account: props.Env.Account,
			Region:  jsii.String("us-west-2"),
		},
	})

	sourceStack := stacks.NewMultiRegionS3CrrKmsCmkSourceStack(stage, jsii.String("MultiRegionS3CrrKmsCmkSource"), &stacks.MultiRegionS3CrrKmsCmkSourceStackProps{
		StackProps: awscdk.StackProps{
			Env: &awscdk.Environment{
				Account: props.Env.Account,
				Region:  jsii.String("us-west-1"),
			},
		},
		TargetBucket:                targetStack.TargetBucket,
		TargetKeyIdSsmParameterName: targetStack.TargetKeyIdSsmParameterName,
		TargetRegion:                targetStack.Region,
	})

	sourceStack.Stack.AddDependency(targetStack.Stack, nil)

	return stage
}
